Preliminary implementation of Period Chart

stevebaranski opened the pull request:
### Description
Fixes #11063, which calls for the introduction of chart styling for period data, with an emphasis on the overlay of Visitors / Views data. 

To test:

- Checkout the branch, confirm that it builds & tests pass.
- For a site, navigate to Stats, and select a given period (i.e., Days, Weeks, Months, Years). 
- Please note that at the moment, the chart is the same regardless of period or dimension (Views, Visitors, Likes, Comments).

Update release notes:

- [X] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### Exhibits

![11063pr](https://user-images.githubusercontent.com/221062/53986264-81699180-40d2-11e9-9d8b-18c5aca0ef6a.gif)

_iPad - Portrait_
<img width="878" alt="11063pr_ipadportrait" src="https://user-images.githubusercontent.com/221062/53986211-626aff80-40d2-11e9-9357-bece93442141.png">

_iPad - Landscape_
<img width="1109" alt="11063pr_ipadlandscape" src="https://user-images.githubusercontent.com/221062/53986210-626aff80-40d2-11e9-99c2-66e45cd3289b.png">

_iPhone SE_
<img width="504" alt="11063pr_iphonese" src="https://user-images.githubusercontent.com/221062/53986212-626aff80-40d2-11e9-836c-c72a72665752.png">

stevebaranski commented:
@ScoutHarris would you be willing to take a look?

@SylvesterWilmott I have included a few screenshots depicting the chart in the proper place. While some of the filter bar wrapping / clipping observed in #11204 remains, hopefully this will give you a sense of things.

In the case of the overlaid bars, I had to adjust the opacity of the highlight so that the Visitors bar was visible. This washes out a bit. I'm open to any suggestions on how I could make this better.

frosty commented:
> In the case of the overlaid bars, I had to adjust the opacity of the highlight so that the Visitors bar was visible. This washes out a bit. I'm open to any suggestions on how I could make this better.

I wonder if perhaps we need a `secondaryBarHighlightColor`, and have them both fully opaque? I can't see how we'll get the desired colors using translucency. 🤔

SylvesterWilmott commented:
@planarvoid How was this implemented on Android?

stevebaranski commented:
> I wonder if perhaps we need a `secondaryBarHighlightColor`, and have them both fully opaque? I can't see how we'll get the desired colors using translucency. 

@frosty thanks for this suggestion! I've observed that a single highlight appears to be higher in the z-order than both bars. I had not considered the notion of two highlights. I will give this prompt attention on Monday. I have updated the milestone associated with the PR to reflect this.

frosty commented:
@stevebaranski I'm not sure if I'm using the right terminology – the marker is the translucent background that stretches to the top? That one appears to be behind both bars, is that correct? Is that different from the single highlight?

stevebaranski commented:
@frosty 

> the marker is the translucent background that stretches to the top?

Yes. It is translucent, and extends from the top of the bar to the top of the chart. In the framework, it's a custom view (hence, the rotation issue you observed before). Other uses might include a "bubble", or even a label with the specific data value.

> Is that different from the single highlight?

Yes. The highlight uses the existing bar as a mask. In the case where Views/Visitors are presented together, highlighting is enabled for the View bar, but disabled for Visitors. Unfortunately, however, the highlight occludes the Visitors bar.

@SylvesterWilmott was right to suggest that I reach out to @planarvoid; I've done that via Slack, and will consider his input on Monday.

Depending on what that change entails, I'll see if it makes sense to add commits to this PR, or to consider a change in a subsequent PR. As it stands, the PR can still be reviewed, but the aesthetics may improve.

ScoutHarris commented on WordPress/Classes/ViewRelated/Stats/Period Stats/Overview/OverviewCell.swift:
```diff
@@ -97,4 +131,26 @@ private extension OverviewCell {
+        }
+    }
+
+    func configureChartViewIfNeeded() {
```
I'm a little confused by this method name, specifically the `IfNeeded` part. Is the chart ever _not_ needed on the Overview? There isn't a condition within the the method that indicates that scenario.

stevebaranski replied:
This is valid feedback. I've pushed a commit ([`3779c0a`](https://github.com/wordpress-mobile/WordPress-iOS/pull/11210/commits/3779c0a6d9e499c94047fb9fa85bb9449b2c3073)) that removes the `IfNeeded` suffix.

ScoutHarris commented:
Hey @stevebaranski . I left just one minor comment/question. I'll wait to give a final review until you decide if any additional changes are coming.

planarvoid commented:
hey @stevebaranski, the code handling this functionality is in `org.wordpress.android.ui.stats.refresh.lists.sections.viewholders.BarChartViewHolder`. 
What I do is I have 3 data sets. A base data set with `Views` which has `blue_wordpress` color and `orange_active` as highlight color. A second data set (`Visitors`) which has `blue_dark` color and `orange_dark_active` as highlight color (`highlightAlpha` is set to `255` in both cases). The last data set is what I call a Highlight data set in which I manually set the values to max value of the chart (so it fills the whole column. This data set has a default color set to `transparent` so it's invisible by default and its highlight color is `orange_active` with alpha set to `51`. 
Be careful when you add the data sets to the `BarData` because the order matters. I tried loads of variations and in the end I add first the default `Views`, then the `Visitors` and then the `Highlight`. 
The real magic is how I handle highlighting. The default highlight doesn't work with multiple datasets (especially since one is invisible). I'm setting my custom `OnChartValueSelectedListener` which calls a method `highlightColumn` which does this:
```
if (hasOverlappingColumns) {
            val high = Highlight(index.toFloat(), 0, 0)
            val high2 = Highlight(index.toFloat(), 1, 1)
            val high3 = Highlight(index.toFloat(), 2, 2)
            high.dataIndex = index
            high2.dataIndex = index
            high3.dataIndex = index
            highlightValues(arrayOf(high3, high, high2))
        } else {
            val high = Highlight(index.toFloat(), 0, 0)
            val high2 = Highlight(index.toFloat(), 1, 1)
            high.dataIndex = index
            high2.dataIndex = index
            highlightValues(arrayOf(high2, high))
        }
```

As you see, I'm setting the highlight manually to the data sets. Be careful, the order matters! It took me ages to get it right. This works with datasets set to [Views, Visitors, Highlight]. Let me know if this helps!

SylvesterWilmott commented:
@stevebaranski this all looks so great! Thanks for spending the time to perfect the charting.

In landscape mode and tablet will we increase the number of bars we are showing?

stevebaranski commented:
Hey @SylvesterWilmott - thanks for this feedback. 

> In landscape mode and tablet will we increase the number of bars we are showing?

At the moment, the stubbed data has 10 values. Once we start to integrate with the "live" data set, I'd like to work with you to optimize the appearance for each device. Potentially, we can control the number of rows that appear, or fine tune the bar width, etc. If you want to keep an eye on the relevant issues, #11067, #11068, #11069 are the ones I have in mind.

stevebaranski commented:
@ScoutHarris I have addressed your feedback with an additional commit. I have also created #11230 to capture the styling changes discussed herein. I will tackle that next, but would like to perform that work in a separate PR.

Do you mind revisiting this review?
